_posts/2022-12-31-arrow-adbc.md

```diff
@@ -0,0 +1,217 @@
+Roughly speaking, when an application executes a query with these APIs:
+
+<figure style="text-align: center;">
+  <img src="{{ site.baseurl }}/img/ADBCFlow1.svg" width="90%" class="img-responsive" alt="A diagram showing the query execution flow.">
+  <figcaption>The query execution flow.</figcaption>
+</figure>
+
+1. The application submits a SQL query via the JDBC/ODBC API.
+2. The query is passed on to the driver.
+3. The driver translates the query to a database-specific protocol and sends it to the database.
+4. The database executes the query and returns the result set in a database-specific format.
+5. The driver translates the result format into the JDBC/ODBC API.
+6. The application iterates over the result rows using the JDBC/ODBC API.
+
+When columnar data comes into play, however, problems arise.
+JDBC is a row-oriented API, and while ODBC can support columnar data, the type system and data representation is not a perfect match with Arrow.
+So generally, columnar data must be converted to rows between steps 5 and 6, spending resources without performing "useful" work.
+
+This mismatch is problematic for columnar database systems, such as ClickHouse, Dremio, DuckDB, and Google BigQuery.
+On the client side, tools such as Apache Spark and pandas would be better off getting columnar data directly, skipping that conversion.
+Otherwise, they're leaving performance on the table.
+At the same time, that conversion isn't always avoidable.
+Row-oriented database systems like PostgreSQL aren't going away, and these clients will still want to consume data from them.
+
+Developers have a few options:
+
+- *Just use JDBC/ODBC*.
+  These standards are here to stay, and it makes sense for databases to support them for applications that want them.
+  But when both the database and the application are columnar, that means converting data into rows for JDBC/ODBC, only for the client to convert them right back into columns!
+  Performance suffers, and developers have to spend time implementing the conversions.
+- *Use JDBC/ODBC-to-Arrow conversion libraries*.
+  Libraries like [Turbodbc][turbodbc] and [arrow-jdbc][arrow-jdbc] handle row to columnar conversions for clients.
+  But this doesn't fundamentally solve the problem.
+  Unnecessary data conversions are still required.
+- *Use vendor-specific protocols*.
+  For some databases, applications can use a database-specific protocol or SDK to directly get Arrow data.
+  For example, applications could use Dremio via [Arrow Flight SQL][flight-sql].
+  But client applications that want to use multiple database vendors would need to integrate with each of them.
+  (Look at all the [connectors](https://trino.io/docs/current/connector.html) that Trino implements.)
+  And databases like PostgreSQL don't offer an option supporting Arrow in the first place.
+
+So in the status quo, clients must choose between either tedious integration work or leaving performance on the table.
```

Comment:
```suggestion
As is, clients must choose between either tedious integration work or leaving performance on the table. We can make this better.
```

I think this could be punchier as a section ender.

Reply:
Thanks! Updated the preview.